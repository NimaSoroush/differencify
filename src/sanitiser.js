import check from 'check-types';
import type from 'type-detect';
import logger from './utils/logger';
import { globalConfig, testConfig } from './config/defaultConfigs';

const logError = (name, wrongType, correctType) => logger.error(
  `Invalid argument ${name} with type ${wrongType} been passed. Argument should be ${correctType}`,
);

const checkProperty = (obj, property, checkType) => {
  if (!obj) {
    return false;
  }
  const hasProperty = Object.prototype.hasOwnProperty.call(obj, property);
  if (!check[checkType](obj[property]) && hasProperty) {
    logError(property, type(obj[property]), checkType);
    return false;
  }
  return hasProperty;
};

const sanitiseTestConfiguration = (conf, testId) => {
  const configuration = {};
  configuration.chain = checkProperty(conf, 'chain', 'boolean')
    ? conf.chain
    : testConfig.chain;
  configuration.testNameProvided = checkProperty(conf, 'testName', 'string');
  configuration.testName = configuration.testNameProvided
    ? conf.testName
    : testConfig.testName;
  if (testId !== undefined) {
    configuration.testId = testId;
  }
  configuration.isUpdate = (process.env.update && process.env.update === 'true')
    ? process.env.update
    : testConfig.isUpdate;
  return configuration;
};

const sanitiseGlobalConfiguration = (conf) => {
  const configuration = {};
  configuration.debug = checkProperty(conf, 'debug', 'boolean')
    ? conf.debug
    : globalConfig.debug;

  configuration.imageSnapshotPathProvided = checkProperty(conf, 'imageSnapshotPath', 'string');
  configuration.imageSnapshotPath = configuration.imageSnapshotPathProvided
    ? conf.imageSnapshotPath
    : globalConfig.imageSnapshotPath;
  configuration.saveDifferencifiedImage = checkProperty(conf, 'saveDifferencifiedImage', 'boolean')
    ? conf.saveDifferencifiedImage
    : globalConfig.saveDifferencifiedImage;
  configuration.saveCurrentImage = checkProperty(conf, 'saveCurrentImage', 'boolean')
    ? conf.saveCurrentImage
    : globalConfig.saveCurrentImage;
  configuration.mismatchThreshold = checkProperty(conf, 'mismatchThreshold', 'number')
    ? conf.mismatchThreshold
    : globalConfig.mismatchThreshold;

  return configuration;
};

export { sanitiseTestConfiguration, sanitiseGlobalConfiguration };
