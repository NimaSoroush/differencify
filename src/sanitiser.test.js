import { sanitiseGlobalConfiguration, sanitiseTestConfiguration } from './sanitiser';

describe('sanitiser', () => {
  describe('sanitise global configuration', () => {
    it('sanitise if no config provided', () => {
      const configuration = sanitiseGlobalConfiguration({});
      expect(configuration).toMatchSnapshot();
    });
    it('sanitise if screenshots config provided', () => {
      const configuration = sanitiseGlobalConfiguration({ screenshots: './somewhere' });
      expect(configuration).toMatchSnapshot();
    });
  });
  describe('sanitise test configuration', () => {
    it('sanitise if no config provided', () => {
      const configuration = sanitiseTestConfiguration({}, 1);
      expect(configuration).toMatchSnapshot();
    });
    it('sanitise if some config provided', () => {
      const configuration = sanitiseTestConfiguration({ newWindow: true }, 1);
      expect(configuration).toMatchSnapshot();
    });
  });
});
