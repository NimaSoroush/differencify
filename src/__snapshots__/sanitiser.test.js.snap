// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`sanitiser sanitise global configuration sanitise if no config provided 1`] = `
Object {
  "debug": false,
  "isUpdate": false,
  "mismatchThreshold": 0.01,
  "puppeteer": Object {
    "args": Array [],
    "dumpio": false,
    "executablePath": undefined,
    "headless": true,
    "ignoreHTTPSErrors": false,
    "slowMo": 0,
    "timeout": 30000,
  },
  "saveDifferencifiedImage": true,
  "screenshots": "./screenshots",
  "testReports": "./differencify_report",
}
`;

exports[`sanitiser sanitise global configuration sanitise if screenshots config provided 1`] = `
Object {
  "debug": false,
  "isUpdate": false,
  "mismatchThreshold": 0.01,
  "puppeteer": Object {
    "args": Array [],
    "dumpio": false,
    "executablePath": undefined,
    "headless": true,
    "ignoreHTTPSErrors": false,
    "slowMo": 0,
    "timeout": 30000,
  },
  "saveDifferencifiedImage": true,
  "screenshots": "./somewhere",
  "testReports": "./differencify_report",
}
`;
